### PR TITLE
[Docs] Add new_task example; improve: feature store engines descriptions, log_artifact, function handler

### PR DESCRIPTION
**Pull request by jillnogold**
[ML-4087](https://jira.iguazeng.com/browse/ML-4087), [ML-3669](https://jira.iguazeng.com/browse/ML-3669), [ML-4920](https://jira.iguazeng.com/browse/ML-4920), [ML-4093](https://jira.iguazeng.com/browse/ML-4093)

### CODE REVIEW

File: docs/runtimes/functions.md [hunk 1]
```diff
@@ -1,27 +1,73 @@
 (Functions)=
-# Functions
+# Functions 
 
-All the executions in MLRun are based on **Serverless functions**. The functions allow specifying code and 
-all the operational aspects (image, required packages, [cpu/mem/gpu resources](../runtimes/configuring-job-resources.html#cpu-gpu-and-memory-limits-for-user-jobs), [storage](../runtimes/function-storage.html), environment, etc.). 
-The [different function runtimes](../concepts/functions-overview.html) take care of automatically transforming the code and spec to fully 
-managed and elastic services over Kubernetes, which saves significant operational overhead, 
-addresses scalability and reduces infrastructure costs.
+All the executions in MLRun are based on serverless functions. Functions are essentially Python objects that know how to run locally or on a Kubernetes cluster.
```
Comment (theSaarco) [inline docs/runtimes/functions.md:4]:
```suggestion
All the executions in MLRun are based on serverless functions. Functions are essentially Python code that can be executed locally or on a Kubernetes cluster. 
```

I don't like the "objects that know how to run..." idiom. The entire idea is that the code doesn't need to know anything about where it runs.

File: docs/runtimes/functions.md [hunk 2]
```diff
@@ -1,27 +1,73 @@
...
 
-Function objects are all inclusive (code, spec, API, and metadata definitions), which allows placing them 
-in a shared and versioned function market place. This means that different members of the team can produce or 
-consume functions. Each function is versioned and stored in the MLRun database with a unique hash code, 
+Function objects are all inclusive, containing the code and all the operational aspects: ({ref}`image <images-usage>`, required packages, 
+{ref}`pod resource congifuration (replicas, CPU/GPU/memory limits, volumes, Spot vs. On-demand nodes, pod priority, node affinity) <configuring-job-resources>`, 
+[storage](../runtimes/function-storage.html), environment, metadata definitions, etc.). Each function is versioned and stored in the MLRun database with a unique hash code, 
 and gets a new hash code upon changes.
 
+When you create a function, you specify its name, which appears in the MLRun UI, as well as the Kubernetes pod.
```
Comment (theSaarco) [inline docs/runtimes/functions.md:39]: This sentence is not clear - it may be understood from it that you specify its k8s pod, which not only is not correct, it also negates what's written above.
I assume you meant that the the k8s pod used to execute the code appears in the MLRun UI, but why is this relevant here? This sentence deals with what happens when you create a function.

File: docs/runtimes/functions.md [hunk 3]
```diff
@@ -1,27 +1,73 @@
...
+
+Depending on the job type, the actual python function 
+can be included in the code as a {py:func}`~mlrun.handler`. You can also use the handler to log results and artifacts, see {ref}`decorators-and-auto-logging`.
```
Comment (theSaarco) [inline docs/runtimes/functions.md:42]: This sentence is also confusing. As the `handler` you point to here is the MLRun decorator. I think a better explanation would be something like:

> MLRun functionality such as logging results and artifacts can be added to existing code using the {py:func}`~mlrun.handler` decorator. You can also use MLRun auto-logging to automatically track machine learning processes such as training a model and record results and artifacts seamlessly while they execute. See {ref}`decorators-and-auto-logging`.

File: docs/runtimes/functions.md [hunk 4]
```diff
@@ -1,27 +1,73 @@
...
+
+## Function hub
+
+Since function objects are all-inclusive (code, spec, API, and metadata definitions), they can be stored in, 
+and reused from, a shared and versioned function hub. This means that different members of the team can produce or 
+consume functions.
```
Comment (theSaarco) [inline docs/runtimes/functions.md:48]: Partially true - it's definitely not easy to share new functions through the hub, and using a custom, private hub, is currently not easily supported. It is easy to consume functions from the hub, maybe we should focus on that...
If we want to emphasize the sharing of functions, maybe it's better to focus on how many users can share the same MLRun project and get access to objects associated with it.

**Review by theSaarco — COMMENTED** [review]
Added some comments on the `runtimes/functions.md` file.

File: docs/runtimes/functions.md (hunk 3 above)
Comment (alonmr) [reply docs/runtimes/functions.md:42]: Worth to note that the handler decorator is now effective by default since the introduction of packagers.

File: docs/runtimes/images.md [hunk 5]
```diff
@@ -20,12 +21,12 @@ Every release of MLRun includes several images for different usages. All images
 
 The images are:
 
-`mlrun/mlrun`: An MLRun image includes preinstalled OpenMPI. Useful as a base image for simple jobs.
-`mlrun/mlrun-gpu`: The same as `mlrun/mlrun` but for GPUs, including Open MPI. 
-`mlrun/ml-base`: The image for file acquisition, compression, dask jobs, simple training jobs and other utilities. Like `mlrun/mlrun` with the addition of Miniconda and other python packages.
-`mlrun/jupyter`: An image with Jupyter giving a playground to use MLRun in the open source. Built on top of jupyter/scipy-notebook, with the addition of MLRun and several demos and examples.
-`mlrun/mlrun-api`: The image used for running the MLRun API.
-`mlrun/mlrun-ui`: The image used for running the MLRun UI.
+- `mlrun/mlrun`: An MLRun image includes preinstalled OpenMPI. Useful as a base image for simple jobs.
```
Comment (alonmr) [inline docs/runtimes/images.md:24]:
```suggestion
- `mlrun/mlrun`: An MLRun image includes preinstalled OpenMPI and other ML packages. Useful as a base image for simple jobs.
```

File: docs/runtimes/images.md [hunk 6]
```diff
@@ -20,12 +21,12 @@ Every release of MLRun includes several images for different usages. All images
...
+- `mlrun/mlrun-gpu`: The same as `mlrun/mlrun` but for GPUs, including Open MPI. 
+- `mlrun/ml-base`: The image for file acquisition, compression, dask jobs, simple training jobs and other utilities. Like `mlrun/mlrun` with the addition of Miniconda and other python packages.
```
Comment (alonmr) [inline docs/runtimes/images.md:26]: Not sure how file acquisition or compression are specifically related to this image. Also, mlrun/mlrun is not based on miniconda as well.
```suggestion
- `mlrun/ml-base`: Similar to `mlrun/mlrun` with the addition of Dask and other python packages.
```

File: docs/runtimes/images.md [hunk 7]
```diff
@@ -20,12 +21,12 @@ Every release of MLRun includes several images for different usages. All images
...
+- `mlrun/jupyter`: An image with Jupyter giving a playground to use MLRun in the open source. Built on top of jupyter/scipy-notebook, with the addition of MLRun and several demos and examples.
+- `mlrun/mlrun-api`: The image used for running the MLRun API.
+- `mlrun/mlrun-ui`: The image used for running the MLRun UI.
```
Comment (alonmr) [inline docs/runtimes/images.md:29]: These 2 are not related to runtimes images. Don't think they should be here.

**Review by alonmr — COMMENTED** [review]
A few comments

File: docs/runtimes/create-and-use-functions.ipynb [hunk 8]
```diff
@@ -5,6 +5,7 @@
    "id": "8d3e3a9c",
```
Comment (alonmr) [inline docs/runtimes/create-and-use-functions.ipynb:5]: <blockquote>MLRun adds, by default, "nuclio-"  </blockquote><p>I'm assuming this is only relevant to nuclio functions</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlrun/mlrun/pull/4756/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: docs/runtimes/functions.md (hunk 1 above)
Comment (jillnogold) [reply docs/runtimes/functions.md:4]: updated

File: docs/runtimes/functions.md (hunk 3 above)
Comment (jillnogold) [reply docs/runtimes/functions.md:42]: @alonmr  what do you mean "effective by default"? Enabled?

File: docs/runtimes/functions.md (hunk 4 above)
Comment (jillnogold) [reply docs/runtimes/functions.md:48]: ...that sentence has been there for ages. Good to get it fixed!

File: docs/feature-store/feature-sets.md [hunk 9]
```diff
@@ -36,9 +36,10 @@ Create a {py:class}`~mlrun.feature_store.FeatureSet` with the base definitions:
    are determined by the relations field if it exists, and otherwise by the entities.
 * **timestamp_key** &mdash; (optional) Used for specifying the time field when joining by time.
 * **engine** &mdash; The processing engine type:
-   - Spark
-   - pandas
-   - storey. Default. (Some advanced functionalities are in the Beta state.)
+   - Spark &mdash; Good for simple batch transformations
```
[the remark this answers is not in the document]
Comment (jillnogold) [reply docs/feature-store/feature-sets.md:39]: updated

File: docs/projects/run-build-deploy.md [hunk 10]
```diff
@@ -84,9 +89,31 @@ run2 = project.run_function("train", inputs={"dataset": run1.outputs["data"]})
 run2.artifact('confusion-matrix').show()
 ```
 
-```{admonition} Run/simulate functions locally: 
-Functions can also run and be debugged locally by using the `local` runtime or by setting the `local=True` 
-parameter in the {py:meth}`~mlrun.runtimes.BaseRuntime.run` method (for batch functions).
+Example of `new_function` with `new_task`:
+
+```
+import mlrun
```
[the remark this answers is not in the document]
Comment (jillnogold) [reply docs/projects/run-build-deploy.md:95]: Point taken, updated. Will go thru docs

File: docs/runtimes/images.md (hunk 5 above)
Comment (jillnogold) [reply docs/runtimes/images.md:24]: updated

File: docs/runtimes/images.md (hunk 6 above)
Comment (jillnogold) [reply docs/runtimes/images.md:26]: updated

File: docs/runtimes/create-and-use-functions.ipynb (hunk 8 above)
Comment (jillnogold) [reply docs/runtimes/create-and-use-functions.ipynb:5]: updated

File: docs/runtimes/images.md (hunk 7 above)
Comment (jillnogold) [reply docs/runtimes/images.md:29]: deleted, and changed the heading name to **MLRun runtime images**

File: docs/runtimes/functions.md (hunk 3 above)
Comment (alonmr) [reply docs/runtimes/functions.md]: https://github.com/mlrun/mlrun/blob/development/mlrun/package/__init__.py#L51

Comment (jillnogold) [reply docs/runtimes/functions.md]: updated to 'enabled by default'

File: docs/projects/run-build-deploy.md [hunk 11]
```diff
@@ -84,10 +89,31 @@ run2 = project.run_function("train", inputs={"dataset": run1.outputs["data"]})
 run2.artifact('confusion-matrix').show()
 ```
 
-```{admonition} Run/simulate functions locally: 
-Functions can also run and be debugged locally by using the `local` runtime or by setting the `local=True` 
-parameter in the {py:meth}`~mlrun.runtimes.BaseRuntime.run` method (for batch functions).
+Example of `new_function` with `new_task`:
```
Comment (theSaarco) [inline docs/projects/run-build-deploy.md:92]: Generally speaking, we want to shift from using the global `new_function`/`run_function` method to project methods. Since we're making changes here, this may be a good opportunity to do it.
The only function that we don't have a project equivalent for is the `new_task` function.

File: docs/runtimes/create-and-use-functions.ipynb [hunk 12]
```diff
@@ -153,7 +107,7 @@
    "id": "bb9cebf9",
    "metadata": {},
    "source": [
-    "- **name:** Name of your MLRun function within the given project. This is displayed in the MLRun UI, as well as the Kubernetes pod.\n",
+    "- **name:** Name of your MLRun function within the given project. This is displayed in the MLRun UI and the Kubernetes pod.\n",
```
Comment (theSaarco) [inline docs/runtimes/create-and-use-functions.ipynb:110]:
```suggestion
    "- **name:** Name of your MLRun function within the given project. This is displayed in the MLRun UI along with the Kubernetes pod name.\n",
```

File: docs/runtimes/create-and-use-functions.ipynb [hunk 13]
```diff
@@ -163,7 +117,7 @@
     "\n",
     "- **kind:** Runtime the MLRun function uses. See [**Kinds of functions (runtimes)**](../concepts/functions-overview.html) for the list of supported batch and real-time runtimes.\n",
     "\n",
-    "- **handler:** Default function handler to invoke (e.g. a Python function within your script). This handler can also be overridden when executing the function.\n",
+    "- **handler:** Default function handler to invoke (e.g. a Python function within your code). This handler can also be overridden when executing the function. For example, `project.run_function(\"func1\", handler=\"f2\")` executes the `f2 code`, even if the default handler is `f1`.\n",
```
Comment (theSaarco) [inline docs/runtimes/create-and-use-functions.ipynb:120]:
```suggestion
    "- **handler:** Default function handler to invoke (e.g. a Python function within your code). This handler can also be overridden when executing the function. For example, `project.run_function(\"func1\", handler=\"f2\")` executes the `f2` Python function, even if the default handler is `f1`.\n",
```

File: docs/runtimes/functions.md [hunk 14]
```diff
@@ -1,27 +1,73 @@
...
-in a shared and versioned function market place. This means that different members of the team can produce or 
-consume functions. Each function is versioned and stored in the MLRun database with a unique hash code, 
+Function objects are all inclusive, containing the code and all the operational aspects: ({ref}`image <images-usage>`, required packages, 
+{ref}`pod resource congifuration (replicas, CPU/GPU/memory limits, volumes, Spot vs. On-demand nodes, pod priority, node affinity) <configuring-job-resources>`, 
+[storage](../runtimes/function-storage.html), environment, metadata definitions, etc.). Each function is versioned and stored in the MLRun 
+database with a unique hash code, 
 and gets a new hash code upon changes.
 
+MLRun functionality such as logging results and artifacts can be added to existing code using the {py:func}`mlrun.handler` decorator, 
+enabled by default.
```
Comment (theSaarco) [inline docs/runtimes/functions.md:26]: Since the handler decorator is now enabled by default (as part of packagers) I would say it's actually a bad practice to use the handler anymore. Instead we should provide details on how to specify these configurations using the `returns` parameter and the input type hints. @guy1992l , @alonmr  - WDYT?

File: docs/store/artifacts.md [hunk 15]
```diff
@@ -3,19 +3,23 @@
 
 An artifact is any data that is produced and/or consumed by functions, jobs, or pipelines. 
 
-Artifacts metadata is stored in the project’s database. The main types of artifacts are:
+Each artifact has a kind that is assigned when the artifact is created, which are also
```
Comment (theSaarco) [inline docs/store/artifacts.md:6]:
```suggestion
There are several types of Artifacts. The type of the Artifact is reflected in the `kind` attribute of each Artifact. These types are also 
```

File: docs/store/artifacts.md [hunk 16]
```diff
@@ -3,19 +3,23 @@
...
 
-- Files — files, directories, images, figures, and plotlines
-- Datasets — any data, such as tables and DataFrames
-- Models — all trained models
+- Files — Files, directories, images, figures, and plotlines
+- Datasets — Any data, such as tables and DataFrames
+- Models — All trained models
 - Feature Store Objects — Feature sets and feature vectors
 
+Artifacts metadata is stored in the project’s database.
```
Comment (theSaarco) [inline docs/store/artifacts.md:15]:
```suggestion
Artifacts metadata is stored in MLRun's database. 
```

File: docs/store/artifacts.md [hunk 17]
```diff
@@ -56,15 +61,15 @@ The default `artifact_path` can be specified at the cluster level, client level,
 You can set the default `artifact_path` for your environment using the {py:func}`~mlrun.set_environment` function.
 
 You can override the default `artifact_path` configuration by setting the `artifact_path` parameter of 
-the {py:func}`~mlrun.set_environment` function. You can use variables in the artifacts path, 
+the {py:func}`~mlrun.projects.MlrunProject.artifact_path` function. You can use variables in the artifacts path,
```
Comment (theSaarco) [inline docs/store/artifacts.md:64]: This is not accurate - by setting the `MlrunProject.artifact_path` property you override the default `artifact_path` configuration **for that specific project**.

**Review by theSaarco — CHANGES_REQUESTED** [review]
Some small changes are needed.

File: docs/runtimes/functions.md (hunk 14 above)
Comment (alonmr) [reply docs/runtimes/functions.md:26]: Agreed

Comment (jillnogold) [reply docs/runtimes/functions.md]: @theSaarco  - updated text, pls check

File: docs/projects/run-build-deploy.md (hunk 11 above)
Comment (jillnogold) [reply docs/projects/run-build-deploy.md]: - replaced 2 instances of  `...as project methods, or as global (mlrun.) methods` with `...as project methods`
- updated new_task example

File: docs/projects/run-build-deploy.md [hunk 18]
```diff
@@ -84,11 +86,32 @@ run2 = project.run_function("train", inputs={"dataset": run1.outputs["data"]})
...
 ```
 
-```{admonition} Run/simulate functions locally: 
-Functions can also run and be debugged locally by using the `local` runtime or by setting the `local=True` 
-parameter in the {py:meth}`~mlrun.runtimes.BaseRuntime.run` method (for batch functions).
+Example with `new_task`:
+
+```python
+import mlrun
+mlrun.get_or_create_project('example-project')
```
Comment (theSaarco) [inline docs/projects/run-build-deploy.md:93]:
```suggestion
project = mlrun.get_or_create_project('example-project')
```

File: docs/projects/run-build-deploy.md [hunk 19]
```diff
@@ -84,11 +86,32 @@ run2 = project.run_function("train", inputs={"dataset": run1.outputs["data"]})
...
+from mlrun import RunTemplate, new_task, mlconf
+from os import path
+artifact_path = path.join(mlconf.artifact_path, '{{run.uid}}')
+def handler(context, param, model_names):
+    context.logger.info("Running handler")
+    context.set_label('category', 'tests')
+    for model_name, file_name in model_names:
+        context.log_artifact(model_name, body=param.encode(), local_path=file_name)
+----
+func = mlrun.set_function("my-func", kind="job", image="mlrun/mlrun")
```
Comment (theSaarco) [inline docs/projects/run-build-deploy.md:105]:
```suggestion
func = project.set_function("my-func", kind="job", image="mlrun/mlrun")
```

File: docs/projects/run-build-deploy.md [hunk 20]
```diff
@@ -84,11 +86,32 @@ run2 = project.run_function("train", inputs={"dataset": run1.outputs["data"]})
...
+func.save()
+---
+task = new_task(name='mytask', handler=handler, artifact_path=artifact_path, project='project-name')
+run_object = project.run_function("my-func", local=True, base_task=task)
+run_object.uid()
```
Comment (theSaarco) [inline docs/projects/run-build-deploy.md:110]: Why is this line needed?

File: docs/runtimes/create-and-use-functions.ipynb [hunk 21]
```diff
@@ -163,7 +117,7 @@
     "\n",
     "- **kind:** Runtime the MLRun function uses. See [**Kinds of functions (runtimes)**](../concepts/functions-overview.html) for the list of supported batch and real-time runtimes.\n",
     "\n",
-    "- **handler:** Default function handler to invoke (e.g. a Python function within your script). This handler can also be overridden when executing the function.\n",
+    "- **handler:** Default function handler to invoke (e.g. a Python function within your code). This handler can also be overridden when executing the function. For example, `project.run_function(\"func1\", handler=\"f2\")` executes the `f2 Python function`, even if the default handler is `f1`.\n",
```
Comment (theSaarco) [inline docs/runtimes/create-and-use-functions.ipynb:120]:
```suggestion
    "- **handler:** Default function handler to invoke (e.g. a Python function within your code). This handler can also be overridden when executing the function. For example, `project.run_function(\"func1\", handler=\"f2\")` executes the `f2` Python function, even if the default handler is `f1`.\n",
```

File: docs/store/artifacts.md [hunk 22]
```diff
@@ -3,19 +3,24 @@
...
 An artifact is any data that is produced and/or consumed by functions, jobs, or pipelines. 
 
-Artifacts metadata is stored in the project’s database. The main types of artifacts are:
-
-- Files — files, directories, images, figures, and plotlines
-- Datasets — any data, such as tables and DataFrames
-- Models — all trained models
+There are several types of Artifacts. The type of the Artifact is reflected in the 
+`kind` attribute of each Artifact. These types are also 
+used in the for grouping the [artifacts in the UI](#viewing-artifacts).
```
Comment (theSaarco) [inline docs/store/artifacts.md:8]:
```suggestion
used for grouping the [artifacts in the UI](#viewing-artifacts). 
```

File: docs/store/artifacts.md [hunk 23]
```diff
@@ -56,15 +62,15 @@ The default `artifact_path` can be specified at the cluster level, client level,
 You can set the default `artifact_path` for your environment using the {py:func}`~mlrun.set_environment` function.
 
 You can override the default `artifact_path` configuration by setting the `artifact_path` parameter of 
-the {py:func}`~mlrun.set_environment` function. You can use variables in the artifacts path, 
+the {py:func}`~mlrun.projects.MlrunProject.artifact_path` function. You can use variables in the artifacts path,
```
Comment (theSaarco) [inline docs/store/artifacts.md:65]:
```suggestion
the {py:class}`~mlrun.projects.MlrunProject` object, setting the artifact path for objects belonging to that project. You can use variables in the artifacts path, 
```

**Review by theSaarco — CHANGES_REQUESTED** [review]
Some additional changes needed.

File: docs/projects/run-build-deploy.md (hunk 20 above)
Comment (jillnogold) [reply docs/projects/run-build-deploy.md]: deleted run_object.uid()

**Review by theSaarco — APPROVED** [review]
LGTM

File: docs/runtimes/functions.md (hunk 2 above)
Comment (jillnogold) [reply docs/runtimes/functions.md]: deleted
